test/suite/test_hs17.py

```diff
@@ -0,0 +1,108 @@
+# software under copyright law.
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+# EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+# MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
+# IN NO EVENT SHALL THE AUTHORS BE LIABLE FOR ANY CLAIM, DAMAGES OR
+# OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE,
+# ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
+# OTHER DEALINGS IN THE SOFTWARE.
+
+import time, wiredtiger, wttest
+
+def timestamp_str(t):
+    return '%x' % t
+
+# test_hs17.py
+class test_hs17(wttest.WiredTigerTestCase):
+    conn_config = 'cache_size=5MB'
+    session_config = 'isolation=snapshot'
+
+    def test_hs17(self):
+        uri = 'table:test_hs17'
+        self.session.create(uri, 'key_format=S,value_format=S')
+        session2 = self.setUpSessionOpen(self.conn)
+        cursor = self.session.open_cursor(uri)
+        cursor2 = session2.open_cursor(uri)
+
+        value1 = 'a' * 500
+        value2 = 'b' * 500
+        value3 = 'c' * 500
+        value4 = 'd' * 500
+        value5 = 'e' * 500
+
+        # Insert an update at timestamp 5
+        self.session.begin_transaction()
+        cursor[str(0)] = value1
+        self.session.commit_transaction('commit_timestamp=' + timestamp_str(5))
+
+        # Insert another update at timestamp 10
+        self.session.begin_transaction()
+        cursor[str(0)] = value2
+        self.session.commit_transaction('commit_timestamp=' + timestamp_str(10))
+
+        # Insert a bunch of other contents to trigger eviction
+        for i in range(1, 1000):
+            self.session.begin_transaction()
+            cursor[str(i)] = value3
+            self.session.commit_transaction()
+
+        # Start a long running transaction which could see update 0.
+        session2.begin_transaction()
+
+        cursor2.set_key(str(0))
+
+        # Assert we can see a value
+        self.assertEqual(cursor2.search(), 0)
+        session2.breakpoint()
+        val1 = cursor2.get_value()
+        cursor2.reset()
+
+        # Commit an update without a timestamp on our original key
+        self.session.begin_transaction()
+        cursor[str(0)] = value4
+        self.session.commit_transaction()
+
+        # Commit an update with timestamp 15
+        self.session.begin_transaction()
+        cursor[str(0)] = value5
+        self.session.commit_transaction('commit_timestamp=' + timestamp_str(15))
+
+        # Insert a bunch of other contents to trigger eviction
+        for i in range(1, 1000):
+            self.session.begin_transaction()
+            cursor[str(i)] = value3
+            self.session.commit_transaction()
+
+        cursor2.set_key(str(0))
+        # Given the bug exists this will return WT_NOTFOUND.
+        session2.breakpoint()
+        self.assertEqual(cursor2.search(), 0)
+        val2 = cursor2.get_value()
+        self.assertEqual(val1, val2)
```

Comment:
I think you can get the same value here because you triggered an eviction before you do the mixed mode operation. If you only trigger one eviction after you commit value 5, I believe you will still get not found as value1 and value2 will not be inserted to history store as they are behind a mixed mode operation.

Reply:
Also have you tried to make the transaction in session2 to read with timestamp5? I think it will return value2 instead of value1 in the current implementation.

Reply:
The test isn't as significant to review I have about 5 different tests that I'll push up. I'm more interested in reviewing hs_rec.c